### PR TITLE
payout reversion keeps funds in treasury

### PR DESCRIPTION
**Pull request by mejango**
5 changes:

- if _pay reverts, add funds back to balance of paying project. This requires removing the gas-optimizing block when paying through the current terminal because cant try/catch calls to internal fns.
- send the referrer projectId in first 32 bytes of metadata, even in _pay for fee.
- distributePayoutsOf takes a _metadata param instead of a _memo param. _metadata can be used for specifying referring client's projectId.
- useAllowanceOf takes a _metadata param which can be used for specifying referring client's projectId.
- new IJBFeeHoldingTerminal interface exposing an addToBalance call with a flag giving callers option to refund held fees.

### CODE REVIEW

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 1]
```diff
@@ -0,0 +1,1646 @@
...
+      // Trigger the allocator's `allocate` function.
+      // If this terminal's token is ETH, send it in msg.value.
+      try
+        _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data)
+      {} catch {
+        // Set the net payout amount to 0 to signal the reversion.
+        netPayoutAmount = 0;
+
+        // Add undistributed amount back to project's balance.
+        store.recordAddedBalanceFor(_projectId, _amount);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1211]: In ERC20 terminals this is exploitable a malicious user can:
1. Create a project
2. Deposit some amount of an ERC20
3. Create an Allocator that always reverts when `allocate` gets called
4. Distribute all funds to the allocator
5. The terminal calls `_beforeTransferTo` which sets an allowance, but the call reverts, the balance gets added back to the project
6. The malicious user calls a method on their allocator to use the allowance and transfer it to them
7. Go back to step 4

We could change `_beforeTransferTo` to increase the allowance and add another method (ex. `_cancelTransferTo`) to decrease it, in the catch we then decrease it with `_amount` to get back to the original allowance.

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 2]
```diff
@@ -0,0 +1,1646 @@
...
+            token,
+            '',
+            _projectMetadata
+          )
+        {} catch {
+          // Set the net payout amount to 0 to signal the reversion.
+          netPayoutAmount = 0;
+
+          // Add undistributed amount back to project's balance.
+          store.recordAddedBalanceFor(_projectId, _amount);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1254]: See [this comment](https://github.com/jbx-protocol/juice-contracts-v3/pull/17/files#r1103702332)

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 3]
```diff
@@ -0,0 +1,1646 @@
...
+            _split.preferClaimed,
+            '',
+            _projectMetadata
+          )
+        {} catch {
+          // Set the net payout amount to 0 to signal the reversion.
+          netPayoutAmount = 0;
+
+          // Add undistributed amount back to project's balance.
+          store.recordAddedBalanceFor(_projectId, _amount);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1273]: See [this comment](https://github.com/jbx-protocol/juice-contracts-v3/pull/17/files#r1103702332)

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 4]
```diff
@@ -0,0 +1,1646 @@
...
+        token,
+        _beneficiary,
+        0,
+        false,
+        '',
+        _projectMetadata
+      )
+    {} catch {
+      // Add fee amount back to project's balance.
+      store.recordAddedBalanceFor(_from, _amount);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1372]: See [this comment](https://github.com/jbx-protocol/juice-contracts-v3/pull/17/files#r1103702332)

File: contracts/JBERC20PaymentTerminal3_1.sol [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  /** 
+    @notice
+    Logic to be triggered before transferring tokens from this terminal.
+
+    @param _to The address to which the transfer is going.
+    @param _amount The amount of the transfer, as a fixed point number with the same number of decimals as this terminal.
+  */
+  function _beforeTransferTo(address _to, uint256 _amount) internal override {
+    IERC20(token).safeApprove(_to, _amount);
```
Comment (xBA5ED) [inline contracts/JBERC20PaymentTerminal3_1.sol:112]: Should we replace this with `safeIncreaseAllowance`? `safeApprove` has been [deprecated](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5a00628ed3d6ce3154cee4d2cc93fad920e8ea30/contracts/token/ERC20/utils/SafeERC20.sol#L31) and reverts if a contract has a remaining allowance when it gets assigned a new one.

Comment (mejango) [reply contracts/JBERC20PaymentTerminal3_1.sol:112]: ok by me. though maybe we do also want to create a _decreaseAllowance in the case of revert. meh?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol (hunk 1 above)
Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1211]: yes. agree with fix. worth doing.

File: contracts/JBERC20PaymentTerminal3_1.sol (hunk 5 above)
Comment (mejango) [reply contracts/JBERC20PaymentTerminal3_1.sol:112]: saw other comment. lfg.

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol (hunk 1 above)
Comment (viraj124) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1226]: nice catch

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 6]
```diff
@@ -0,0 +1,1671 @@
...
+        decimals,
+        _projectId,
+        _group,
+        _split
+      );
+
+      // Trigger the allocator's `allocate` function.
+      // If this terminal's token is ETH, send it in msg.value.
+      try
+        _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data)
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1217]: I'm wondering if we shouldn't test for interface id instead, something like

```
if(ERC165Checker.supportsInterface(_split.allocator, IJBAllocator.interfaceId) _split.allocator.allocate(...)
else {
    _cancelTransferTo
    (...)
}
```

rn, if the allocator is either an eoa and a contract with a fallback, the call will not revert (and fund lost)

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 7]
```diff
@@ -0,0 +1,1678 @@
...
+        store.recordAddedBalanceFor(_projectId, _amount);
+      }
+
+      // Otherwise, if a project is specified, make a payment to it.
+    } else if (_split.projectId != 0) {
+      // Get a reference to the Juicebox terminal being used.
+      IJBPaymentTerminal _terminal = directory.primaryTerminalOf(_split.projectId, token);
+
+      // The project must have a terminal to send funds to.
+      if (_terminal == IJBPaymentTerminal(address(0))) revert TERMINAL_IN_SPLIT_ZERO_ADDRESS();
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1242]: Should we validate it's really a terminal (erc165) too?

Comment (viraj124) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1242]: I was also thinking of this yesterday, only reason I had not to was the scope i.e whether this should be the responsibility of the project owner, but if we are doing it for allocator then you're right we should do it for terminal too

Comment (drgorillamd) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1242]: nvm, when calling setPrimaryTerminal in the directory, a bunch of static calls are made to the terminal -> these would revert if it's not a valid terminal

**Review by drgorillamd — APPROVED** [review]

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol [hunk 8]
```diff
@@ -0,0 +1,1681 @@
...
+
+      // If the terminal is set as feeless, this distribution is not eligible for a fee.
+      if (_feeDiscount == JBConstants.MAX_FEE_DISCOUNT || isFeelessAddress[address(_terminal)])
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1250]:
```suggestion
      if (address(_terminal) == address(this) ||
          _feeDiscount == JBConstants.MAX_FEE_DISCOUNT ||
          isFeelessAddress[address(_terminal)])
```

Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1250]: nice

Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1.sol:1250]: is this already in an `else`?

**Pull request merged**